### PR TITLE
Seq optimization

### PR DESCRIPTION
**Pull request by KhalilCK**
We replaced certain List methods with Seq package.
The sanity check passed without errors using make test and the benchmarking using the following flags:
 

`./goblint ../bench/coreutils/ls_comb.c --set pre.cppflags[+] "--std=gnu89" --disable ana.base.context.non-ptr --disable ana.int.def_exc --disable sem.unknown_function.spawn --set ana.thread.domain plain --enable exp.earlyglobs --set ana.base.privatization none --set pre.cppflags[+] -DGOBLINT_NO_BSEARCH --set pre.cppflags[+] -DGOBLINT_NO_ASSERT --set result json-messages --set ana.activated "['base', 'mallocWrapper']" --set ana.ctx_insens[+] base --set ana.ctx_insens[+] mallocWrapper --enable dbg.timing.enabled --enable ana.int.interval --enable ana.int.interval_threshold_widening`

On my local machine the Seq implementation took 820.1s.
We will choose between the Seq and iter packages, which one is better based on the results of the benchmarking on the cluster that you can find in the Iter PR https://github.com/goblint/analyzer/pull/964.

### CODE REVIEW

File: make.sh [hunk 1]
```diff
@@ -11,6 +11,14 @@ opam_setup() {
   opam switch -y create . --deps-only ocaml-base-compiler.4.14.0 --locked
 }
 
+opam_setup_flambda() {
+  set -x
+  opam init -y -a --bare $SANDBOXING # sandboxing is disabled in travis and docker
+  opam update
+  # Note: the `--update-invariant` option is needed for replacement of the previous ocaml compiler switch invariant of Goblint
+  opam switch -y create . --deps-only ocaml-variants.4.14.0+options ocaml-option-flambda --locked --update-invariant
+}
```
Comment (jerhard) [inline make.sh:20]: Are these changes necessary or already subsumed by the merged flambda PR?

File: src/analyses/base.ml [hunk 2]
```diff
@@ -85,7 +83,7 @@ struct
    * Helpers
    **************************************************************************)
 
-  let is_privglob v = GobConfig.get_bool "annotation.int.privglobs" && v.vglob
+   let is_privglob v = get_bool "annotation.int.privglobs" && v.vglob
```
Comment (jerhard) [inline src/analyses/base.ml:86]: Is this change in indentation intentional?

File: src/analyses/base.ml [hunk 3]
```diff
@@ -546,7 +544,8 @@ struct
     done;
     (* Return the list of elements that have been visited. *)
     if M.tracing then M.traceu "reachability" "All reachable vars: %a\n" AD.pretty !visited;
-    List.map AD.singleton (AD.elements !visited)
+     List.of_seq @@ (Seq.map AD.singleton) @@ List.to_seq (AD.elements !visited)
```
Comment (jerhard) [inline src/analyses/base.ml:547]: Since there is only operation (`List.map`) that is replaced, can one hope that this change actually can have a positive performance impact?

File: src/analyses/base.ml [hunk 4]
```diff
@@ -855,7 +854,7 @@ struct
           | [] -> assert false
           | (e1, e2) :: eqs ->
             let eqs_for_all_mem e = List.for_all (fun (e1, e2) -> CilType.Exp.(equal e1 e || equal e2 e)) eqs in
-            let eqs_map_remove e = List.map (fun (e1, e2) -> if CilType.Exp.equal e1 e then e2 else e1) eqs in
+            let eqs_map_remove e = List.of_seq @@ Seq.map (fun (e1, e2) -> if CilType.Exp.equal e1 e then e2 else e1) @@ List.to_seq eqs in
```
Comment (jerhard) [inline src/analyses/base.ml:857]: Same question as above, as there is only  a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 5]
```diff
@@ -867,7 +866,7 @@ struct
           let* eqs = split exp in
           let* (e, es) = find_common eqs in
           let v = eval_rv a gs st e in (* value of common exp *)
-          let vs = List.map (eval_rv a gs st) es in (* values of other sides *)
+          let vs = List.of_seq @@ (Seq.map (eval_rv a gs st)) @@ List.to_seq es in (* values of other sides *)
```
Comment (jerhard) [inline src/analyses/base.ml:869]: Same question as above, as there is only a List.map that is replaced.

File: src/analyses/base.ml [hunk 6]
```diff
@@ -1224,13 +1223,13 @@ struct
     | Q.EvalLength e -> begin
         match eval_rv_address (Analyses.ask_of_ctx ctx) ctx.global ctx.local e with
         | `Address a ->
-          let slen = List.map String.length (AD.to_string a) in
+          let slen = List.of_seq @@ (Seq.map String.length)@@ List.to_seq (AD.to_string a) in
```
Comment (jerhard) [inline src/analyses/base.ml:1226]: Same question as above, as there is only a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 7]
```diff
@@ -1224,13 +1223,13 @@ struct
...
           let lenOf = function
             | TArray (_, l, _) -> (try Some (lenOfArray l) with LenOfArray -> None)
             | _ -> None
           in
-          let alen = List.filter_map (fun v -> lenOf v.vtype) (AD.to_var_may a) in
-          let d = List.fold_left ID.join (ID.bot_of (Cilfacade.ptrdiff_ikind ())) (List.map (ID.of_int (Cilfacade.ptrdiff_ikind ()) %BI.of_int) (slen @ alen)) in
+          let alen = List.of_seq @@ Seq.filter_map (fun v -> lenOf v.vtype)@@ List.to_seq (AD.to_var_may a) in
```
Comment (jerhard) [inline src/analyses/base.ml:1231]: Same question as above, as there is only a `List.filter_map` that is replaced. 
Maybe, if the `alen` is only used once, it would not have to be converted to a `list` here?

File: src/analyses/base.ml [hunk 8]
```diff
@@ -1347,7 +1346,7 @@ struct
     | `Union _ ->
       begin
         let vars_in_partitioning = VD.affecting_vars value in
-        let dep_new = List.fold_left (fun dep var -> add_one_dep x var dep) st.deps vars_in_partitioning in
+        let dep_new = Seq.fold_left (fun dep var -> add_one_dep x var dep) st.deps @@ List.to_seq vars_in_partitioning in
```
Comment (jerhard) [inline src/analyses/base.ml:1349]: Same question as above, as there is only a `List.fold_left` that is replaced.

File: src/analyses/base.ml [hunk 9]
```diff
@@ -1403,7 +1402,7 @@ struct
         if M.tracing then M.tracel "set" ~var:firstvar "update_one_addr: returning: '%a' is a function type \n" d_type x.vtype;
         st
       end else
-      if get_bool "exp.globs_are_top" then begin
+        if get_bool "exp.globs_are_top" then begin
```
Comment (jerhard) [inline src/analyses/base.ml:1405]: Is this change in indentation intentional?

File: src/analyses/base.ml [hunk 10]
```diff
@@ -2268,7 +2267,7 @@ struct
     (* First we create a variable-initvalue pair for each variable *)
     let init_var v = (AD.from_var v, v.vtype, VD.init_value ~varAttr:v.vattr v.vtype) in
     (* Apply it to all the locals and then assign them all *)
-    let inits = List.map init_var f.slocals in
+    let inits = List.of_seq @@ Seq.map init_var @@ List.to_seq f.slocals in
```
Comment (jerhard) [inline src/analyses/base.ml:2270]: Same question as above, as there is only a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 11]
```diff
@@ -2331,7 +2330,7 @@ struct
         | `Address a -> AD.remove NullPtr a
         | _ -> AD.empty ()
       in
-      List.map mpt exps
+      List.of_seq @@ Seq.map mpt @@ List.to_seq exps
```
Comment (jerhard) [inline src/analyses/base.ml:2333]: Same question as above, as there is only a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 12]
```diff
@@ -2349,16 +2348,16 @@ struct
...
     let invalids = invalidate_exp exps in
     let is_fav_addr x =
       List.exists BaseUtil.is_excluded_from_invalidation (AD.to_var_may x)
     in
     let invalids' = List.filter (fun (x,_,_) -> not (is_fav_addr x)) invalids in
     if M.tracing && exps <> [] then (
-      let addrs = List.map (Tuple3.first) invalids' in
-      let vs = List.map (Tuple3.third) invalids' in
+      let addrs = List.of_seq @@ Seq.map (Tuple3.first) @@ List.to_seq invalids' in
+      let vs = List.of_seq @@ Seq.map (Tuple3.third) @@ List.to_seq invalids' in
```
Comment (jerhard) [inline src/analyses/base.ml:2360]: Same question as above, as there is only a `List.map` that is replaced in each line.

File: src/analyses/base.ml [hunk 13]
```diff
@@ -2367,7 +2366,7 @@ struct
   let make_entry ?(thread=false) (ctx:(D.t, G.t, C.t, V.t) Analyses.ctx) fundec args: D.t =
     let st: store = ctx.local in
     (* Evaluate the arguments. *)
-    let vals = List.map (eval_rv (Analyses.ask_of_ctx ctx) ctx.global st) args in
+    let vals = List.of_seq @@ Seq.map (eval_rv (Analyses.ask_of_ctx ctx) ctx.global st) @@ List.to_seq args in
```
Comment (jerhard) [inline src/analyses/base.ml:2369]: Same question as above, as there is only a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 14]
```diff
@@ -2419,7 +2418,7 @@ struct
         let args =
           match arg with
           | Some x -> [x]
-          | None -> List.map (fun x -> MyCFG.unknown_exp) fd.sformals
+          | None -> List.of_seq @@ Seq.map (fun x -> MyCFG.unknown_exp) @@ List.to_seq fd.sformals
```
Comment (jerhard) [inline src/analyses/base.ml:2421]: Same question as above, as there is only a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 15]
```diff
@@ -2514,7 +2513,7 @@ struct
       (addr, AD.get_type addr)
     in
     let forks = forkfun ctx lv f args in
-    if M.tracing then if not (List.is_empty forks) then M.tracel "spawn" "Base.special %s: spawning functions %a\n" f.vname (d_list "," d_varinfo) (List.map BatTuple.Tuple3.second forks);
+    if M.tracing then if not (List.is_empty forks) then M.tracel "spawn" "Base.special %s: spawning functions %a\n" f.vname (d_list "," d_varinfo) (List.of_seq @@ Seq.map BatTuple.Tuple3.second @@ List.to_seq forks);
```
Comment (jerhard) [inline src/analyses/base.ml:2516]: Same question as above, as there is only a `List.map` that is replaced.

File: src/analyses/base.ml [hunk 16]
```diff
@@ -2649,7 +2648,7 @@ struct
         | `Address ret_a ->
           begin match eval_rv (Analyses.ask_of_ctx ctx) gs st id with
             | `Thread a ->
-              let v = List.fold VD.join (VD.bot ()) (List.map (fun x -> G.thread (ctx.global (V.thread x))) (ValueDomain.Threads.elements a)) in
+              let v = List.fold VD.join (VD.bot ()) (List.of_seq @@ Seq.map (fun x -> G.thread (ctx.global (V.thread x))) @@ List.to_seq (ValueDomain.Threads.elements a)) in
```
Comment (jerhard) [inline src/analyses/base.ml:2651]: Same question as above, as there is only a `List.map` that is replaced.

File: src/cdomains/intDomain.ml [hunk 17]
```diff
@@ -609,20 +662,35 @@ struct
...
+      let config_value = Some (get_bool "ana.int.interval_threshold_widening") in
+      interval_threshold_widening_ref := config_value; !interval_threshold_widening_ref
+    else
+      !interval_threshold_widening_ref
+
+  let extract_from_option (x: float_precision option): float_precision =
+    match x with
+    | None -> failwith "Config option not read"
+    | Some x -> x
+*)
```
Comment (jerhard) [inline src/cdomains/intDomain.ml:679]: Please remove the blocks of commented code (here and below).

**Review by jerhard — COMMENTED** [review]
Hey, thanks for your pull request! 
Good to see that you could find places where it looks like it makes sense to introduce `Seq`!

I left some comments, mainly about the case when only one operation on a list is replaced by an operation on `Seq`, and some conversion back and forth (from and to lists) is done. I would assume that in such cases, there would be no performance benefit of doing this, as still the same number of allocations has to be done to create the result list -- only in `List.of_seq` instead of, say `List.map`.

**Comment by KhalilCK** [comment]
As we discussed I removed all the Seq that are performed on only one single List operation.

**Comment by sim642** [comment]
Closing in favor of #975.
